### ISSUE TITLE
Design & Implement support for instrumentation-based metrics

### ISSUE THREAD

**Issue by blue42u** [HPCToolkit/hpctoolkit#491]
#### Background
Some GPU architectures (specifically Intel) don't support PC sampling and don't have an API for that kind of tracing. On these platforms the only mechanism to obtain fine-grained GPU performance data is through instrumentation. Instrumentation-based metrics are fundamentally different from the sampling-based metrics we use everywhere else: sampling-based metrics are attributed everywhere (applicable) in the application code, while instrumentation-based metrics are attributed to the non-application instrumentation probes.

From Prof2's perspective, the primary need is that metrics attributed to the probes are re-attributed to the applicable instructions. I originally thought this was in the same vein as https://github.com/HPCToolkit/hpctoolkit/issues/440, however after a Slack conversation with @aarontcopal2 I have concluded that the differences are fundamental enough to consider it a separate front. The key difference here is that #440 is about reconstructing missing calling Contexts, this is about reconstructing missing metric values.

As per the usual with my Prof2 designs, the goal here is to cleanly solve the more general problem of instrumentation-based metrics, and then apply them to Intel GPU support in particular.

#### Proposed Design
Since instrumentation is applied at Run-time, it makes sense to me to save data on the instrumentation then instead of reconstructing it later with post-mortem binary analysis. This information is metric-dependent (and obviously binary-dependent), and additional scaling factors may need to be incorporated for proper re-attribution.

The proposed design is to add a new "instrumentation" section to `.hpcrun` with the appropriate tables. Metrics listed in the metric table point to the probe -> instruction mappings and to the list of factors to apply. For sampling-based metrics the former pointer is `NULL`, `NULL` in the latter pointer just means all factors are `1`.

This new section is read and parsed by Hpcrun4, and all metric re-attributed is handled there without involving the Prof2 core.

Proposed additions to the `.hpcrun` format:
```
*.hpcrun:
...
| Metric Description {                                 |
|   ...                                                |
|   0, or Instrumentation Header offset (inst_ptr)     |
|   0, or Instrumentation Factors offset (factors_ptr) |
| }                                                    |
...
|--- Instrumentation section ----------------------|
| Instrumentation Header (see below)               |
| ...                                              |
| Instrumentation Table (see below)                |
| ...                                              |
| Instrumentation Factors (see below)              |
| ...                                              |
...


Instrumentation Header:
|----------------------------------------------------|
| Number of Instrumented Load Modules (num_mods)     |
| Instrumented Load Module Header 0 {                |
|   Index of Load Module in loadmap section (mod_id) |
|   Offset of Instrumentation Table (table_ptr)      |
| }                                                  |
| Instrumented Load Module Header 1                  |
| ...                                                |
| Instrumented Load Module Header (num_mods-1)       |
|----------------------------------------------------|

Instrumentation Table:
|--------------------------------------------------------------------|
| Number of Probes (num_probes)                                      |
| Probe 0 {                                                          |
|   IP in load module (mod_id) of Probe (probe_ip)                   |
|   Number of Covered Instructions (num_insts)                       |
|   IP in load module (mod_id) of Covered Instruction 0 (inst_ip[0]) |
|   IP in load module (mod_id) of Covered Instruction 1 (inst_ip[1]) |
|   ...                                                              |
|   IP in load module (mod_id) of Covered Instruction (num_insts-1)  |
| }                                                                  |
| Probe 1                                                            |
| ...                                                                |
| Probe (num_probes-1)                                               |
|--------------------------------------------------------------------|

Instrumentation Factors:
|----------------------------------------------------------|
| Factors for Instrumented Load Module Header 0 {          |
|   Factors for Probe 0 {                                  |
|     Factor for Covered Instruction 0                     |
|     Factor for Covered Instruction 1                     |  
|     ...                                                  |
|     Factor for Covered Instruction (num_insts-1)         |
|   }                                                      |
|   Factors for Probe 1                                    | 
|   ...                                                    |
|   Factors for Probe (num_probes-1)                       |
| }                                                        |
| Factors for Instrumented Load Module Header 1            |
| ...                                                      |
| Factors for Instrumented Load Module Header (num_mods-1) |
|----------------------------------------------------------|
```

#### Action Items
 - [ ] Finalize the design of a measurement format for instrumentation mappings, to be written by Run and read by Prof2.
 - [ ] Implement the above format in `prof-lean` as part of the sparse-file interface.
 - [ ] Fill this section with appropriate information during runtime instrumentation of Intel GPU binaries.
 - [ ] Update Hpcrun4 to create the appropriate Contexts and re-attribute metric values as they are parsed.

**Comment by aarontcopal2** [comment]
@blue42u, Although I understood the gist of what you are doing, I had some questions/things to clarify:

1. Instrumentation Header keeps tabs of the instrumented binaries, right?
2. According to the current design, it seems like you are recording 'Instrumentation Table' and 'Instrumentation Factors' for each metric. Is this understanding correct?
3. Will/should these data-structures be responsible for holding the metric values as well? Since the 'Instrumentation table' has the load-module ip, it could record the metric value for the corresponding instruction.
4. What is the purpose of 'Instrumentation Factors'? Does this have any use for the Intel instrumentation use case or is this more catered for the general instrumentation use case?

**Comment by blue42u** [comment]
1. Right, the Header's just to associate the data with load modules, only the instrumented binaries need to be listed there.
2. Almost, the Header (and associated Tables) can be shared between metrics if the probes are identical. For instance I'm guessing the Intel SIMD metrics are all probed together, they can share a single Header (and associated Tables). The Factors are similar (although if the Header isn't shared the Factors won't be either).
3. No. Metric values are in the sparse metrics section. There are performance and technical reasons why that shouldn't change.
4. The Factors is motivated by the Intel case but it should generalize well. Say you instrument a basic block (BB) and measure time (latency). When you re-attribute, you don't want to attribute the full latency cost to every instruction in the BB, you want to apportion it across the instructions. That's not an even apportion either since some instructions have predictable latencies (from our Slack conversation). The easiest way to get that right is to list the factors to apply during re-attribution, one for every instruction in the BB.

   As for why Factors are per-metric, consider the case where you measure latency and cache misses with a single per-BB probe. The predictions and applicable instructions will differ between the two metric so the Factors will be different, but the probe is in the same spot and covers the same instructions so the Header (and Tables) can be shared. Separating them saves a little over 1/2 the space for little cost.

**Comment by aarontcopal2** [comment]
Got it. Thanks

**Comment by mxz297** [comment]
I think the design draft should be sufficient for instruction-level instrumentation measurement, but I think instrumentation can also be used to provide other metrics that are not instruction-level. For example, it may be useful to know how many times a function, a basic block, a loop, or a particular control flow edge are executable; such information can be very useful for code layout profile guide optimizations. 

Of course, we are nowhere near integrating PGO into the workflow (I think a hpcpgo command will be very cool). For this context, I think it would be nice to design a format to support instrumentation-based metrics that describe certain structures of code, instead of just individual instructions.

**Comment by jmellorcrummey** [comment]
i second xiaozhu’s comment that we should consider planning for more than instruction-level instrumentation.

consider path profiling:

T. Ball and J. R. Larus, "Efficient path profiling," Proceedings of the 29th Annual IEEE/ACM International Symposium on Microarchitecture. MICRO 29, 1996, pp. 46-57, doi: 10.1109/MICRO.1996.566449.

M. D. Bond and K. S. McKinley, "Practical path profiling for dynamic optimizers," International Symposium on Code Generation and Optimization, 2005, pp. 205-216, doi: 10.1109/CGO.2005.28.

**Comment by blue42u** [comment]
Thanks for the comments, let me try to brainstorm a bit in this direction:

First, some of this is actually much easier (I think) than is maybe obvious. Prof2 is designed around the concept of "context," exactly what *kind* of context (calling, lexical, physical...) is represented is very flexible. I don't foresee any new glaring issues with writing an `hpcpgo` which generates attribution to a purely "physical" context tree, the only additions would be some new Scope::Types (basic block, "physical" loop, ...) and a Finalizer to expand to physical contexts (role-shifted in https://github.com/HPCToolkit/hpctoolkit/pull/493).

Second, putting this into a bit more context, AFAIK metrics fall into one of three categories:
 - "Steady" metrics (like `cycles` or `REALTIME`), where the measurement by `hpcrun` and final attribution by Prof2 are in the same location. Fully supported by Prof2.
 - "Raised" metrics (like GPU register usage), where the measurement happens at a lower context than its final attribution (mostly because `hpcrun` can't actually represent anything that isn't an instruction in its CCT), so Prof2 "raises" the metric up the Context tree. Not fully implemented but I intend to batch the last pieces with https://github.com/HPCToolkit/hpctoolkit/pull/481 (since all metrics need careful review for the taxonomy anyway).
 - "Lowered" metrics, where the measurement happens at a higher context (e.g. basic block) than its final attribution (e.g. instructions), so Prof2 needs to (conceptually) "lower" the metric down the context tree.

This issue and design is intended for "lowered" metrics, which are always instrumentation-based (AFAIK samples are always instruction-level). Since `hpcrun` represents everything in its CCT as offsets in binaries, the proposed design translates offsets to offsets in a metric-specific manner. I believe this is the only reasonable (and conceptually consistent) approach without significant adjustments to `hpcrun`'s CCT structures (which is possible but harder).

Third, using these concepts, the requested generalization (if I understand correctly) is "lowered" metrics where the final attribution isn't instructions. Another way to look at this is as metrics (conceptually) "lowered" down the context tree to instructions and then "raised" up to its destination context, in practice this is exactly a metric marked as "raised" but also with the Instrumentation Tables for "lowering." The proposed design (+ last bits of "raised" metrics) supports this completely.

So, unless I've missed a case somewhere, I think we have a plan for support of all feasible metrics (we know of so far).

**Comment by mxz297** [comment]
I feel that if we want to fully leverage the power of instrumentation, we need to generalize our definition of "context". In my understanding, our current context are all represented with PC for instructions, loops, and functions. This is fine if all our metrics are instruction-level. 

This is semantically wrong for higher level metrics. Let me use a classic example where we want to collect the number of times a function is executed. First of all, I believe in your classification, it belongs to "Lowered" metrics. My overall belief is that this metric CANNOT be lowered. The reason is that the number of times a function is executed is not necessarily equivalent to the number of times the first instruction in the function is executed. 

Consider a highly optimized function where the first instruction is inside a loop. When this function is called once, the first instruction can be executed multiple times. Without a more careful representation, we will not be able to represent function frequency from basic block frequency. A similar example is that the frequency of a taken conditional jump is not the same as the frequency of the target instruction.

I think once we have augmented our representation of the calling context tree, such instrumentation-based metrics become natural to fit in the new tree. Many instrumentaiton-based metrics are collected in a way that we cannot distinguish its calling context. For example, we put a timer start and a timer end at the beginning and end of a basic block to collect block level cycles. For such cases, we just need a separate CFG with metrics attached to each node. I am not sure whether we can nicely merge this separate CFG with our existing calling context tree, but once we have a generalized calling context tree, the metric representation should follow naturally.

**Comment by blue42u** [comment]
Ok, the names and descriptions were poor, sorry for the confusion. Function and basic block frequency are "raised" metrics since `hpcrun` will attribute them to the first instruction and Prof2 must "raise" them to the enclosing function or basic block Context, the "level of measurement" in `hpcrun` is higher than the instruction-level represented in the binary/offset CCT. As you have realized they aren't "lowered" since the final attribution in Prof2 isn't lower than their level of measurement. Something like memory accesses could be if measured in `hpcrun` with basic block instrumentation, in that case though the metrics would be attributed in Prof2 ("lowered") to all memory instructions in the basic block, not just the first instruction. Thus the need for Instrumentation Tables to dictate where the values get "lowered" to.

(I believe "steady," "raised" and "lowered" metrics should cover every kind of metric, collectively they represent every level of measurement and every final attribution. If there was a metric that doesn't fall into this grouping, it would be unrelated to context and thus different enough to require its own data representations. Metric-traces fall into this category.)

I don't think `hpcrun`'s CCT needs to be generalized, every metric has a single level of measurement based on how it was gathered. On the off-chance the same value is gathered different ways, that can be easily represented as multiple metrics (at different levels) with the same name. Your example with frequency isn't in this case since they measure different values: summing function and basic block frequency gives a nonsensical result. Generalizing the CCT nodes is redundant to "level of measurement" markers on the metrics, and has some additional performance, memory, space and manhour cost. It maybe looks a little nicer conceptually but I don't yet see the benefit.

The "level of measurement" markers on the metrics are how I'm planning (have been for a long while) to implement "raised" metrics, I have only just realized I probably didn't write that down anywhere on the Prof2 project board. Whoops.

* * *

Prof2 core couldn't care less what sort of "context" is represented, but it must still form a strict tree. So basic block contexts can't be mixed in with lexical contexts (consider inlined functions, a single inlined call may take multiple blocks and multiple functions may be inlined into a single block. Similar for source lines.) Consider trying to present a basic-block metric in the Viewer, where should it appear? It might cut across multiple lines and inlined function calls, so a basic-block context can't be injected into the context-tree in the Viewer either. Do we provide "alternatives" for a subtree to switch between lexical and CFG/basic-block data? I fiddled with that idea a bit for logical context representations (for Python), it's doable but harder than it seems to make inclusive/exclusive metric costs still have useful meanings (although this case specifically might not hit any of the *really* hard cases).

At the end of the day, Prof2 attributes measurements to context, what kind of "context" is up to the front-end tool. `hpcprof2` uses lexical context for the Viewer, `hpcpgo` could use physical context. Most metrics are instruction-level ("steady") and so work for all types of context, metrics "lowered" to instruction-level will probably work too. "Raised" metrics and other "lowered" metrics might not always, and if the analysis doesn't match the metrics there isn't much we can do to rectify that.

So, IMO, we don't need to touch contexts. This issue discusses the plan for the only kind of metric without a plan yet, "lowered" metrics.
